Hooks / Rules of Hooks (#11)
reactjs/pt-BR.reactjs.org: content/docs/hooks-rules.md

---
id: hooks-rules
title: Regras dos Hooks
permalink: docs/hooks-rules.html
next: hooks-custom.html
prev: hooks-effect.html
---

*Hooks* é um novo recurso adicionado no React 16.8. Eles permitem que você use o estado (`state`) e outras funcionalidades do React, sem precisar escrever uma classe.

Hooks são funções Javascript, mas você precisa seguir duas regras ao utilizá-los. Nós providenciamos um [plugin ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks) para aplicar essas regras automaticamente:

> ### Use Hooks apenas no nível superior {#only-call-hooks-at-the-top-level}

Reviewer: ```suggestion
### Use Hooks Apenas no Nível Superior {#only-call-hooks-at-the-top-level}
```

Author: @jhonmike Ok! Sugestão comitada.

> ### Use Hooks apenas no nível superior {#only-call-hooks-at-the-top-level}

Reviewer 2: ```suggestion
### Use Hooks apenas no nível superior {#use-hooks-apenas-no-nivel-superior}
```

**Não use Hooks dentro de loops, regras condicionais ou funções aninhadas (funçoes dentro de funções).** Em vez disso, sempre use Hooks no nível superior de sua função React. Seguindo essas regras, você garante que os Hooks serão chamados na mesma ordem a cada vez que o componente renderizar. É isso que permite que o React preserve corretamente o estado dos Hooks quando são usados várias chamadas a `useState` e `useEffect` na mesma função. (Se você ficou curioso, iremos explicar isso melhor [abaixo](#explanation).)

> ### Use Hooks apenas dentro de funções do React {#only-call-hooks-from-react-functions}

Reviewer: ```suggestion
### Use Hooks Apenas Dentro de Funções do React {#only-call-hooks-from-react-functions}
```

**Não use Hooks dentro de funções Javascript comuns.** Em vez disso, você pode:

* ✅  Chamar Hooks em componentes React.
* ✅  Chamar Hooks dentro de Hooks Customizados (Nós iremos aprender sobre eles [na próxima página.](/docs/hooks-custom.html)).

Seguindo essas regras, você garante que toda lógica de estado (`state`) no componente seja claramente visível no código fonte.

> ## ESLint Plugin {#eslint-plugin}

Reviewer 2: ```suggestion
Plugin ESLint {#eslint-plugin}
```

Author: why we should inverse?

---
id: regras-dos-hooks
title: Regras dos Hooks
permalink: docs/pt-BR/regras-dos-hooks.html
next: hooks-customizados.html
prev: hooks-feitos.html
---

*Hooks* é um novo recurso adicionado em React 16.8. Eles permitem que você use o `state` e outras funcionalidades do React, sem precisar escrever uma classe.

Hooks são funções Javascript, mas você precisa se preocupar em seguir duas regras quando for usar-los. Nós também providenciamos um [plugin ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks) para aplicar as regras automaticamente:

> ### Use Hooks apenas em nível superior

Author: Ok!

**Não use Hooks dentro de loops, regras condicionais ou funções aninhadas (funçoes dentro de funções).** Invés disso, sempre use Hooks no nível superior de sua função React. Por seguir as regras, você garante que os Hooks serão chamados na mesma ordem a cada vez que o componente renderizar. É isso que permite que o React preserve corretamente o estado dos Hooks quando são usados vários `useState` e `useEffect` na mesma função. (Se você ficou curioso, iremos explicar isso melhor [abaixo](#explanation).)

### Use Hooks apenas em funções React {#only-call-hooks-from-react-functions}

**Não use Hooks em funções javascript regulares.** Invés disso, você pode:

* ✅  Chamar Hooks em componentes React.
* ✅  Chamar Hooks dentro de Hooks Customizados (Nós iremos aprender sobre eles [na próxima página.](/docs/hooks-customizados.html)).

> Por seguir essas regras, você garante que toda lógica de estado (`state`) no componente seja claramente visível no código fonte.

Reviewer 3: ```suggestion
Seguindo essas regra, você garante que toda lógica de estado (`state`) no componente seja claramente visível no código fonte.
```

> ## ESLint Plugin

Reviewer 3: ```suggestion
## ESLint Plugin {#eslint-plugin}
```

> Nós liberamos um plugin ESLint chamado [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks) que aplica as duas regras. Se você desejar pode adicionar este plugin ao seu projeto, dessa forma:

Reviewer 3: ```suggestion
Nós liberamos um plugin ESLint chamado [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks) que aplica essas duas regras. Se você desejar pode adicionar este plugin ao seu projeto, dessa forma:
```

```bash

> <code>npm install eslint-plugin-react-hooks@next</code>

Reviewer 3: ```suggestion
npm install eslint-plugin-react-hooks
```

Reviewer 3: Don't change code blocks

```

> </code>

Reviewer 3: ```suggestion

```

}
```

No futuro, temos a intenção de incluir esse plugin por padrão dentro do Create React App e ferramentas similares.

**Você pode pular para próxima página, onde explica melhor como escrever [seus próprios Hooks](/docs/hooks-custom.html) agora.**. Nessa página continuaremos explicando o motivo por trás dessas regras.

## Explicação {#explanation}

Assim como [aprendemos anteriormente](/docs/hooks-state.html#tip-using-multiple-state-variables), nós podemos usar diversos Hooks (States ou Effects) em um único componente:

```js
function Form() {
  // 1. Use a variável de estado (state) name
  const [name, setName] = useState('Mary');

  // 2. Use um efeito para persistir o formulário
  useEffect(function persistForm() {
    localStorage.setItem('formData', name);
  });

  // 3. Use a variável de estado (state) surname
  const [surname, setSurname] = useState('Poppins');

  // 4. Use um efeito para atualizar o título
  useEffect(function updateTitle() {
    document.title = name + ' ' + surname;
  });

  // ....
}
```

> Agora, como o React sabe qual o estado (`state`) correspondente ao `useState` chamado? A resposta é que o **React depende da ordem em que os Hooks são chamados.** Nosso exemplo funciona porque a ordem de chamada dos Hooks é a mesma sempre que o componente é renderizado:

Reviewer 2: ```suggestion
Mas, como o React sabe qual o estado (`state`) correspondente ao `useState` chamado? A resposta é que o **React depende da ordem em que os Hooks são chamados.** Nosso exemplo funciona porque a ordem de chamada dos Hooks é a mesma sempre que o componente é renderizado:
```

Author: @glaucia86 I desagree with `Mas`, becasue we need enfactize in the moment. So, `Agora` sound is good.

useEffect(updateTitle)     // 🔴  Na ordem era pra ser 3 (mas foi 4). Falha ao substituir o efeito
```

O React não saberia o que retornar na segunda chamada do Hook `useState`. O React esperava que a segunda chamada de Hook nesse componente fosse ao efeito `persistForm`, assim como aconteceu na renderização anterior, mas a ordem foi alterada. A partir daí, toda vez que um Hook for chamado depois daquele que nós pulamos, o próximo também se deslocaria, levando a erros.

**É por isso que os Hooks devem ser chamados no nível superior de nosso componente.** Se nós queremos executar um efeito condicional, nós podemos colocar a condição _**dentro**_ de nosso Hook:

```js
useEffect(function persistForm() {
  // 👍  Legal! Agora não quebramos mais a primeira regra.
  if (name !== '') {
    localStorage.setItem('formData', name);
  }
});
```

**Note que você não precisa se preocupar com esse problema, se você usar a [regra fornecida no plugin do ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks)**. Mas agora você também sabe o *porquê* os Hooks funcionam dessa maneira, e quais os problemas que essas regras previnem.

> ## Próximos passos {#next-steps}

Reviewer: ```suggestion
## Próximos Passos {#next-steps}
```

Reviewer 2: @jhonmike obrigada por sinalizar Jhon

}
```

No futuro, temos a intenção de incluir esse plugin por padrão dentro do Create React App e ferramentas similares.

> **Você pode pular para próxima página, onde explica melhor como escrever [seus próprios Hooks](/docs/hooks-custom.html) agora.**. Nessa página, nós continuaremos explicando o motivo por trás dessas regras.

Reviewer 3: ```suggestion
**Você pode pular para próxima página, onde explica melhor como escrever [seus próprios Hooks](/docs/hooks-custom.html) agora.**. Nessa página continuaremos explicando o motivo por trás dessas regras.
```

## Explicação {#explanation}

> Assim como [aprendemos anteriormente](/docs/hooks-state.html#tip-using-multiple-state-variables), nós podemos usar muitos States ou Effects em um único componente:

Reviewer 3: ```suggestion
Assim como [aprendemos anteriormente](/docs/hooks-state.html#tip-using-multiple-state-variables), nós podemos usar diversos Hooks (States ou Effects) em um único componente:
```

```js
function Form() {

>   // 1. Use o nome da variável de estado

Reviewer 3: ```suggestion
  // 1. Use a variável de estado (state) name
```

>   const [name, setName] = useState('Mary'). 

Reviewer 3: ```suggestion
  const [name, setName] = useState('Mary');
```

  // 2. Use um efeito para persistir o formulário
  useEffect(function persistForm() {
    localStorage.setItem('formData', name);
  });

>   // 3. Use um sobrenome como variável de estado

Reviewer 3: ```suggestion
  // 3. Use a variável de estado (state) surname
```

  const [surname, setSurname] = useState('Poppins');

  // 4. Use um efeito para atualizar o título
  useEffect(function updateTitle() {
    document.title = name + ' ' + surname;
  });

  // ....
}
```

> Agora, como o React sabe qual o estado correspondente ao `useState` chamado? A resposta é que para o **React** depende da ordem em que os Hooks sãos chamados. Nosso exemplo funciona porque a ordem do Hook chamado é a mesma sempre que o componente é renderizado:

Reviewer 3: ```suggestion
Agora, como o React sabe qual o estado (`state`) correspondente ao `useState` chamado? A resposta é que o **React depende da ordem em que os Hooks são chamados.** Nosso exemplo funciona porque a ordem de chamada dos Hooks é a mesma sempre que o componente é renderizado:
```

```js
// ------------

> // Primeiro render

Reviewer 3: ```suggestion
// Primeira renderização
```

// ------------

> useState('Mary')           // 1. Inicializa o variável de nome com 'Mary'

Reviewer 3: ```suggestion
useState('Mary')           // 1. Inicializa a variável de estado (state) name com 'Mary'
```

useEffect(persistForm)     // 2. Adiciona um efeito para persistir o formulário

> useState('Poppins')        // 3. Inicializa a variável sobrenome com 'Poppins'

Reviewer 3: ```suggestion
useState('Poppins')        // 3. Inicializa a variável de estado (state) surname com 'Poppins'
```

useEffect(updateTitle)     // 4. Adiciona um efeito para atualizar o título

// -------------

> // Segundo render

Reviewer 3: ```suggestion
// Segunda renderização
```

// -------------

> useState('Mary')           // 1. Ler o nome da variável (argumento ignorado)

Reviewer 3: ```suggestion
useState('Mary')           // 1. Ler a variável de estado (state) name (argumento ignorado)
```

useEffect(persistForm)     // 2. Substitui o efeito para persistir no formulário

> useState('Poppins')        // 3. Ler a variável sobrenome (argumento ignorado)

Reviewer 3: ```suggestion
useState('Poppins')        // 3.Ler a variável de estado (state) surname (argumento ignorado)
```

useEffect(updateTitle)     // 4. Substitui o efeito que atualiza o título

// ...
```

> Enquanto a ordem dos Hooks chamados é a mesma entre as renderizações, React pode associar um `state` local a cada um deles. Mas o que acontece se colocarmos uma chamada de Hook (por exemplo, o efeito `persistForm`) dentro de uma condição?

Reviewer 3: ```suggestion
Enquanto a ordem dos Hooks chamados for a mesma entre as renderizações, o React pode associar um estado (`state`) local a cada um deles. Mas o que acontece se colocarmos uma chamada de Hook (por exemplo, o efeito `persistForm`) dentro de uma condição?
```

```js
// 🔴 Nós estaremos quebrando a primeira regra por usar um Hook dentro de uma condição

> 

Reviewer 3: ```suggestion

```

if (name !== '') {
  useEffect(function persistForm() {
    localStorage.setItem('formData', name);
  });
}
```

> A condição `name !== ''` é `true` na primeira renderização, então chamamos o Hook dentro da condição. Entretanto, na próxima renderização o usuário pode limpar o formulário, fazendo com que a condição seja `false`. Agora que pulamos este Hook durante a primeira renderização, a ordem das chamadas tornou-se diferente, veja:

Reviewer 3: ```suggestion
A condição `name !== ''` é `true` na primeira renderização, então chamamos o Hook dentro da condição. Entretanto, na próxima renderização o usuário pode limpar o formulário, fazendo com que a condição seja `false`. Agora que pulamos este Hook durante a renderização, a ordem das chamadas dos Hooks foi alterada:
```

```js

> useState('Mary')           // ✅  1. Lê a variável name (argumento é ignorado)

Reviewer 3: ```suggestion
useState('Mary')           // ✅  1. Lê a variável de estado (state) name (argumento é ignorado)
```

// useEffect(persistForm)  // 🔴  Agora, este Hook foi ignorado!

> useState('Poppins')        // 🔴  Na ordem era pra ser 2 (mas foi 3). Falha ao ler a variável sobrenome

Reviewer 3: ```suggestion
useState('Poppins')        // 🔴  Na ordem era pra ser 2 (mas foi 3). Falha ao ler a variável de estado (state) surname
```

useEffect(updateTitle)     // 🔴  Na ordem era pra ser 3 (mas foi 4). Falha ao substituir o efeito
```

> React não saberia o que retornar na segunda renderização na chamada do Hook `useState`. React esperava que a segunda chamada de Hook nesse componente corresponde-se ao efeito `persistForm`, assim como aconteceu na renderização anterior, mas a ordem foi alterada. Nesse ponto, toda vez que um Hook for chamado depois daquele que nós pulamos, o próximo também se deslocaria, levando a erros.

Reviewer 3: ```suggestion
O React não saberia o que retornar na segunda chamada do Hook `useState`. O React esperava que a segunda chamada de Hook nesse componente fosse ao efeito `persistForm`, assim como aconteceu na renderização anterior, mas a ordem foi alterada. A partir daí, toda vez que um Hook for chamado depois daquele que nós pulamos, o próximo também se deslocaria, levando a erros.
```

**É por isso que os Hooks devem ser chamados no nível superior de nosso componente.** Se nós queremos executar um efeito condicional, nós podemos colocar a condição _**dentro**_ de nosso Hook:

```js
useEffect(function persistForm() {
  // 👍  Legal! Agora não quebramos mais a primeira regra.
  if (name !== '') {
    localStorage.setItem('formData', name);
  }
});
```

> **Note que você não precisa se preocupar com esse problema, se você usar a [regra fornecida no plugin do ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks)**. Mas agora você também sabe *porque* Hooks funcionam dessa maneira, e quais os problemas que essas regras previnem.

Reviewer 3: ```suggestion
**Note que você não precisa se preocupar com esse problema, se você usar a [regra fornecida no plugin do ESLint](https://www.npmjs.com/package/eslint-plugin-react-hooks)**. Mas agora você também sabe o *porquê* os Hooks funcionam dessa maneira, e quais os problemas que essas regras previnem.
```

## Próximos passos {#next-steps}

> Finalmente, estamos prontos para aprender sobre como [escrever nossos próprios Hooks](/docs/hooks-custom.html)! Hooks Customizados deixam você combinar Hooks fornecidos pelo React junto suas próprias abstrações, e reusar a lógica do `state` entre diferentes componentes.

Reviewer 3: ```suggestion
Finalmente, estamos prontos para aprender sobre como [escrever nossos próprios Hooks](/docs/hooks-custom.html)! Hooks Customizados permitem você combinar Hooks fornecidos pelo React em suas próprias abstrações, e reusar a lógica do `state` entre diferentes componentes.
```

Author: @WendellAdriel change was submitted
